taiga2/controllers/endpoint.py

```diff
@@ -532,12 +531,14 @@ def create_upload_session_file(uploadMetadata, sid):
         if data_file is None:
             api_error("Unknown taiga ID: " + existing_taiga_id)
 
-        # When adding a reference to an existing Taiga datafile, it is possible
-        # to provide a new set of custom metadata in the form of key/value pairs.
-        # If new custom_metadata is not provided, copy the key/value pairs from the
-        # original file.
         if custom_metadata == None:
             custom_metadata = data_file.custom_metadata
+        else:
+            custom_metadata = (
+                custom_metadata
+                if data_file.custom_metadata == None
+                else {**data_file.custom_metadata, **custom_metadata}
+            )
```

Comment:
I made a change so that data_file can get the custom_metadata from the immediately prior virtual dataset. Additionally, `{**data_file.custom_metadata, **custom_metadata}` allows us to retain the previous virtual files figshare id, while allowing us to still add on to the custom_metadata.